content/blog/updating-to-cpp11/index.Rmd

```diff
@@ -0,0 +1,61 @@
+* [ ] Add intro sentence, e.g. the standard tagline for the package
+* [ ] `usethis::use_tidy_thanks()`
+-->
+
+Over the summer I had the pleasure of working with Jim Hester on the cpp11 package as a tidyverse summer intern. The cpp11 package is a header-only R package that helps R package developers handle R objects with C++ code. Its goals and syntax are similar to the excellent Rcpp package.
+
+During most of my internship, I worked on triaging issue requests, fixing bugs and adding new features to cpp11. Near the end of the summer, I got to work with Jenny Bryan on converting the readxl package from using Rcpp to using cpp11.
+
+Jim has written a [great post](https://cpp11.r-lib.org/articles/converting.html) about converting packages from Rcpp to cpp11 which I heavily referenced during the process . But there were still some challenges I encountered. To help others going through a similar process, I wanted to review the workflows and tools we used to make this easier.
+
+## Where to start?
+ 
+To get started, I followed the initial set-up steps outlined in Jim’s article and then recompiled, to confirm it was successful. Next, I needed to include either the cpp11/R.hpp header or the macros R_NO_REMAP and STRICT_R_HEADERS. I found it easier to include the macros since the header file needs to come before any headers that use Rcpp. As for which file to start with, I’d suggest starting from any header or C++ file that is the most self-contained. Initially, I made the mistake of selecting the most interconnected file in the readxl package! Since this file was central to the package, I would have had to convert the entire package over to cpp11 before the compilation errors would stop. After Jim and Jenny helped me select a more approachable file, the process was more incremental and I successfully maintained my sanity 🙌🏼.
+
+## The nitty gritty
+
+For the nitty gritty details of converting the code, I definitely took advantage of the table in Jim’s post! That comparison table is great for converting between Rcpp and cpp11 classes. Also, I had to keep track of whether an object was writable or readable. This new feature in cpp11 is great since writable vectors are costly because the data must be fully copied, so using readable where appropriate is a good idea. When I was unsure of whether an object should be readable or writable, I would make it readable and then recompile to see if I was correct.
+
+After I finished converting a file, I recompiled, ran my tests and fixed any compilation errors or test failures. This sometimes meant I had to edit other functions in other files but I only edited files enough to fix the failures. If I did too much it was difficult for me to know what’s causing the test failures. Another incremental approach I used was to convert one function in a file, and then test. This approach worked best for large files. I’d move on to the next file or function only after all my tests were passing again.
```

Comment:
```suggestion
After I finished converting a file, I recompiled, ran my tests and fixed any compilation errors or test failures. This sometimes meant I had to edit other functions in other files, but I only edited files enough to fix the failures. If I did too much it was difficult for me to know what was causing the test failures. Another incremental approach I used was to convert one function in a file, and then test. This approach worked best for large files. I’d move on to the next file or function only after all my tests were passing again.
```

Re: "my" tests

I think this wording is ambiguous, as it leaves me wondering if you have special Rcpp --> cpp11 conversation tests or does this mean you ran the existing tests in the package?

Reply:
I think I've made this clearer now but let me know if it still needs work.

Reply:
Yeah, it's good. I've mark as resolved.